Tests/FBLPromisesTests/FBLPromise+RetryTests.m

```diff
@@ -0,0 +1,241 @@
+/**
+ Copyright 2018 Google Inc. All rights reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at:
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import "FBLPromise+Retry.h"
+
+#import <XCTest/XCTest.h>
+
+#import "FBLPromise+Async.h"
+#import "FBLPromise+Catch.h"
+#import "FBLPromise+Testing.h"
+#import "FBLPromise+Then.h"
+#import "FBLPromisesTestHelpers.h"
+
+@interface FBLPromiseRetryTests : XCTestCase
+@end
+
+@implementation FBLPromiseRetryTests
+
+- (void)testPromiseRetryWithDefaultRetryAttemptAfterInitialReject {
+  // Arrange.
+  // Initial attempt count plus retry attempts count.
+  NSUInteger __block count = 1 + FBLPromiseRetryDefaultAttempts;
+
+  // Act.
+  [[[FBLPromise retry:^id {
+    count -= 1;
+    return count == 0 ? @42 : [NSError errorWithDomain:FBLPromiseErrorDomain code:42 userInfo:nil];
+  }] then:^id(NSNumber *value) {
+    XCTAssertEqual(value, @42);
+    return nil;
+  }] catch:^(NSError __unused *_) {
+    XCTFail(@"Promise should not be resolved with error.");
+  }];
+
+  // Assert.
+  XCTAssert(FBLWaitForPromisesWithTimeout(10.0));
+  XCTAssertEqual(count, 0);
+}
+
+- (void)testPromiseRetryNoRetryAttemptOnInitialFulfill {
+  // Arrange.
+  NSUInteger __block count = 1 + FBLPromiseRetryDefaultAttempts;
+
+  // Act.
+  [[[FBLPromise retry:^id {
+    count -= 1;
+    return @42;
+  }] then:^id(NSNumber *value) {
+    XCTAssertEqual(value, @42);
+    return nil;
+  }] catch:^(NSError __unused *_) {
+    XCTFail(@"Promise should not be resolved with error.");
+  }];
+
+  // Assert.
+  XCTAssert(FBLWaitForPromisesWithTimeout(10.0));
+  XCTAssertEqual(count, FBLPromiseRetryDefaultAttempts);
+}
+
+- (void)testPromiseRetryExhaustsAllRetryAttemptsBeforeRejection {
+  // Arrange.
+  NSInteger customAttempts = 3;
+  NSUInteger __block count = 1 + customAttempts;
+  NSError *retryError = [NSError errorWithDomain:FBLPromiseErrorDomain code:42 userInfo:nil];
+
+  // Act.
+  [[[FBLPromise attempts:customAttempts
+                   delay:FBLPromiseRetryDefaultDelay
```

Comment:
Yeah, the need to specify the rest of the args when you only want to customize the attempt count doesn't seem ideal in ObjC :)

Reply:
Good point! Will add two new ObjC APIs for now:
1) [FBLPromise attempts:retry:]
2) [FBLPromise onQueue:attempts:retry:]

Can also pass `FBLPromiseRetryDefaultAttemptsCount` and `FBLPromiseRetryDefaultDelayInterval` to parameters where the default values are sufficient.